pipeline-model-definition/src/test/resources/options/provisioningTimeout.groovy

```diff
@@ -0,0 +1,46 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+pipeline {
+    agent any
+    options {
+        provisioningTimeout()
```

Comment:
I'm not liking adding a new option for this, but maybe my dislike is due to lack of understanding.  Could you comment a bit more on what you're trying to achieve?  Why not have the top level timeout include provisioning by default?

Reply:
The main idea is not to change the current behaviour, just to create a new option for users.

Reply:
Could an option be added to the existing `timeout` option? So you might have something like this:

```groovy
    options {
        timeout(time:10, unit: "MINUTES", includeProvisioning: true
    }
```

One thing that @joseblas 's proposal would provide is more granularity. If I've understood it correctly, there would two distinct timeouts - one for spinning up the agent, the other for the actual build work.

Reply:
Indeed, the idea is for applying the timeout to the whole process: agent provisioning and the actual work. @kshultzCB, adding that option to timeout was also on my mind, could be an option.

Reply:
The point of timeouts is to limit run duration and having the default not include checkout or agent provisioning is unexpected.  

My reflex is to make change the default behavior to include checkout and agent provisioning.  Especially at top level. If this breaks people who had their timeout set to values that didn't include checkout and provisioning, I'm okay with that.  

We could provide a way for checkout or provisioning to have an additional timeout within the larger timeout, but do we really need to?  

Giving people more options isn't always better.

Reply:
What if you don't care about provisioning? You just care about timeout in your tests?
Indeed, @kshultzCB idea about specific timeout for provisioning sounds good to me.

Reply:
But that isn't what this change is about. 
If someone wanted to do a timeout one part of a stage they can already do that using the `timeout` wrapper inside of `steps`.  If they want a timeout on a stage as a whole (including provisioning and checkout), they use the stage's `option`.  

Right now, declarative users cannot get a timeout on their whole pipeline including the provisioning of their top-level agent.  That is what we need done.

Reply:
The existing behavior is a bug not a feature. 😄

Reply:
I do seem benefit in the extra granularity in that it might be able to speed up when something is going wrong with the pipeline.  Many times people will put in a large timeout due to a heavy test, but often times there can be an issue with provisioning hanging relatively quickly so that the tests never even get executed. Unless you're babysitting the pipeline you've pretty much wasted a good hour or so.  In that kind of scenario, a way to only consider provisioning timeout is usefule.

Reply:
I don't see the issue here as being specific to `timeout` or provisioning time, and I don't think the fix should be specific to `timeout`. My interpretation of the core issue is that at the top level, `agent` is evaluated before `options`, but at the stage level, `agent` is evaluated after `options`, which is inconsistent, and users have no way to control the ordering.

The fix that makes the most sense to me would be to add a `beforeAgent` option to `options`, which would default to `false` at the top level and `true` at the stage level. Users who want `timeout` to include agent provisioning time at the top level could use `beforeAgent(true)` in their top-level `options`. This could also be used for any other block-scope step that a user wants to execute at the top level before the agent (or vice versa for stage-level). My main concern with this approach is that maybe there are cases where someone wants `timeout` to execute before `agent` but also to have a block-scope step at the top level that requires an `agent` context, which wouldn't be possible with this fix.

Special-casing `timeout` in top-level options seems weird to me. Changing top-level `options` to execute before top-level `agent` by default is probably off the table because it would break too many Pipelines. Maybe another fix could be a new `beforeAgentOptions` `options` block at the top level, but that seems awkward too.

Reply:
> I don't see the issue here as being specific to `timeout` or provisioning time, and I don't think the fix should be specific to `timeout`.
...
> Special-casing timeout in top-level options seems weird to me. 

I agree with this.  

> The fix that makes the most sense to me would be to add a `beforeAgent` option to...

Why would we want `options` processing to default to a different behavior at the top-level ? 

> My main concern with this approach is that maybe there are cases where someone wants timeout to execute before agent but also to have a block-scope step at the top level that requires an agent context, which wouldn't be possible with this fix.

What would that be currently?  If we want to add functionality in the future, that is a separate change. Unless we have a strong, concrete example of where this breaks an existing option we should change this to be consistent.  

Let me run some tests.